pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarServiceNameResolver.java

```diff
@@ -96,15 +131,143 @@ public void updateServiceUrl(String serviceUrl) throws InvalidServiceURL {
                 throw new InvalidServiceURL(e);
             }
         }
-        this.addressList = addresses;
+        this.allAddressList = addresses;
+        this.allAddressSet = new HashSet<>(addresses);
```

Comment:
I think you should use `CopyOnWriteArraySet` instead of `HashSet` to ensure thread safety.